python_modules/dagster/dagster/_core/execution/calculate_data_time.py

```diff
@@ -0,0 +1,125 @@
+from datetime import datetime, timezone
+from typing import AbstractSet, Mapping, Optional, Tuple
+
+from ..definitions.events import AssetKey
+from ..instance import DagsterInstance
+
+ROOT_DATA_TAG = ".dagster/root_data_ids"
+
+
+def get_upstream_materialization_times_for_key(
+    instance: DagsterInstance,
+    asset_key: AssetKey,
+    upstream_asset_key_mapping: Mapping[str, AbstractSet[str]],
+) -> Mapping[str, Tuple[Optional[int], Optional[datetime]]]:
+    """
+    Recursive method to derive the timestamps of the materializations of upstream assets
+    which were used to produce a given materialization.
+
+        * For an asset with no parents, this value is just the timestamp of the given
+        materialization.
+
+        * For an asset with a single parent (which is a root), we find the most recent
+        materialization of that parent which happened /before/ the given materialization, as this
+        is the timestamp of the data which was used to produce the given materialization.
+
+        * For an asset with a single parent (which is not a root), we find the most recent
+        materialization of the parent which happened /before/ the given materialization, then find
+        the data times of that event. This process continues recursively up to the root. If an
+        asset has multiple parents, it is possible to have consumed different timestamps of root
+        assets from different parent paths. In these cases, we take the maximum timestamp for each
+        key.
+
+    To make this process more performant, whenever we calculate these values for a given asset
+    materialization, we cache this information in the DagsterInstance database by tagging the
+    event with the timestamps of its upstream materializations. Before calculating any value, we
+    check if we've already stored it, and if so we can exit early.
+    """
+    from dagster._core.events import DagsterEventType
+    from dagster._core.storage.event_log.base import EventRecordsFilter
+
+    def _get_root_keys(asset_key):
+        def _recursive(key_str):
+            upstream_key_strs = upstream_asset_key_mapping[key_str]
+            if not upstream_key_strs:
+                return {key_str}
+            return set().union(
+                *(_recursive(upstream_key_str) for upstream_key_str in upstream_key_strs)
+            )
+
+        return _recursive(asset_key.to_user_string())
+
+    def _get_upstream_materialization_times(record):
+
```

Comment:
nitpick: remove blank line at top of method. per: https://elementl-workspace.slack.com/archives/C03AFEEDDNX/p1666630503242289?thread_ts=1666629931.217449&cid=C03AFEEDDNX.

also, type annotations would be helpful here